src/lib/deployments/database/index.js

```diff
@@ -155,20 +155,63 @@ export function createConnection(config) {
  * @param {Object} conn An existing database connection.
  * @param {String} name Name for the new database.
  */
-export function createDatabase(conn, name) {
-  return conn.raw(`CREATE DATABASE ${name}`);
+export async function ensureDatabase(conn, name) {
+  const results = await conn.raw(
+    `SELECT EXISTS(SELECT datname FROM pg_catalog.pg_database WHERE datname = '${name}')`
+  );
+  const exists = first(results.rows).exists;
+  if (!exists) {
+    await conn.raw(`CREATE DATABASE ${name}`);
+  }
+}
+
+/*
+ * Create a new user in the database if it does not already exist.
+ * @param {Object} conn An existing database connection.
+ * @param {String} name Name for the new user.
+ * @param {String} password The users new password.
+ */
+export async function ensureUser(conn, user, password) {
+  // Check if user exists
+  const result = await conn.raw(
+    `SELECT EXISTS(SELECT usename FROM pg_user WHERE usename = '${user}')`
+  );
+  const exists = first(result.rows).exists;
+  if (!exists) {
+    await conn.raw(
+      `CREATE USER ${user} WITH LOGIN NOSUPERUSER NOCREATEDB NOCREATEROLE NOINHERIT NOREPLICATION CONNECTION LIMIT -1 ENCRYPTED PASSWORD '${password}';`
+    );
+  }
 }
 
 /*
- * Create a schema with new user/role.
+ * Create a new schema if it does not already exist.
+ * @param {Object} conn An existing database connection.
+ * @param {String} schema The name of the desired schema.
+ * @param {String} user The user authorized to this schema.
+ */
+export async function ensureSchema(conn, schema, user) {
+  // Check if schema exists
+  const result = await conn.raw(
+    `SELECT EXISTS(SELECT schema_name FROM information_schema.schemata WHERE schema_name = '${schema}')`
+  );
+  const exists = first(result.rows).exists;
+  if (!exists) {
+    // Create a new schema, granting the new user access.
+    await conn.raw(`CREATE SCHEMA ${schema} AUTHORIZATION ${user}`);
+  }
+}
```

Comment:
This is racey -- two processes could check it, not find it and both try to create it.
```suggestion
  // Create a new schema, granting the new user access.
  await conn.raw(`CREATE SCHEMA IF NOT EXISTS ${schema} AUTHORIZATION ${user}`);
}
```

Reply:
Could we do the user and database creation as `IF NOT EXISTS` too?

Reply:
It doesn't look like CREATE USER or DATABASE have the `IF NOT EXISTS` variant.

Perhaps the best approach is to try and create, but catch and ignore "already exists" errors?

Reply:
https://stackoverflow.com/questions/8092086/create-postgresql-role-user-if-it-doesnt-exist

```sql
DO
$do$
BEGIN
   IF NOT EXISTS (
      SELECT FROM pg_catalog.pg_roles  -- SELECT list can be empty for this
      WHERE  rolname = 'my_user') THEN

      CREATE ROLE my_user LOGIN PASSWORD 'my_password';
   END IF;
END
$do$;```

Reply:
Yea I think in the POC code, I tried to use `IF NOT EXISTS` on the db creation, and realized that wasn't supported, so kind of copy/pasted the same idea to all three of these create statements (db, user, schema).

I think since this is a _creation_ event, we shouldn't be racing to process two duplicate events for the same deployment, unless we somehow double-publish what should be a single event, which should never happen under normal circumstances. 

In the unlikely event of a double-publish of a single event, and the subsequent race to process them, one of these will ultimately fail and be retried. In which case, it should successfully re-run and end up as a no-op, since we've at least made sure to make the event idempotent. 

I think this type of racy behavior would be more likely to happen on update events since they can happen repeatedly for a single deployment. But, those event handlers don't have anything tricky like this since they are only handling helm upgrades, which can already be re-run cleanly.